docs/source/wrappers.rst

```diff
@@ -0,0 +1,487 @@
+providing an interface that should be familiar for sklearn users.
+
+To get started, define your :py:class:`~tensorflow.keras.Model` architecture like you always do,
+but within a callable top-level function (we will call this function ``model_build_fn`` for
+the remained of these docs, but you are free to name it as you wish).
+Then pass this function to :py:class:`.BaseWrapper` in the ``model`` parameter.
+Finally, you can call :py:func:`~scikeras.wrappers.BaseWrapper.fit`
+and :py:func:`~scikeras.wrappers.BaseWrapper.predict`, as with an sklearn
+estimator. The finished code could look something like this:
+
+.. code:: python
+
+    def model_build_fn():
+         model = Model(...)
+         ...
+         return model
+
+    clf = KerasClassifier(model=model_build_fn)
+    clf.fit(X, y)
+    y_pred = clf.predict(X_valid)
+
+Let's see what SciKeras did:
+
+- wraps ``tensorflow.keras.Model`` in an sklearn interface
+- handles encoding and decoding of the target ``y``
+- compiles the :py:class:`~tensorflow.keras.Model` (unless you do it yourself in ``model_build_fn``)
+- makes all ``Keras`` objects serializable so that they can be used with :py:mod:`~sklearn.model_selection`.
+
+SciKeras abstracts away the incompatibilities and data conversions,
+letting you focus on defining your architecture and
+choosing your hyperparameters.
+At the same time, SciKeras is very flexible and can be
+extended with ease, getting out of your way as much as possible.
+
+Initialization
+^^^^^^^^^^^^^^
+
+When you instantiate the :py:class:`.KerasClassifier` or
+:py:class:`.KerasRegressor` instance, only the given arguments are stored.
+These arguments are stored unmodified. For instance, the ``model`` will
+remain uninstantiated. This is to make sure that the arguments you
+pass are not touched afterwards, which makes it possible to clone the
+wrapper instance, for example in a :py:class:`~sklearn.model_selection.GridSearchCV`.
+
+Only when the :py:func:`~scikeras.wrappers.KerasClassifier.fit` or
+:py:func:`~scikeras.wrappers.KerasRegressor.fit` method are called, are the
+different attributes of the wrapper, such as the ``model``, initialized.
+An initialized attribute's name always ends on an underscore; e.g., the
+initialized ``module`` is called ``model_``. (This is the same
+nomenclature as sklearn uses.) Therefore, you always know which
+attributes you set and which ones were created by the wrappers.
+
+Once initialized by calling ``fit``, the wrappers create several attributes,
+documented in the :ref:`scikeras-api` documentation.
+
+Compilation of ``Model``
+^^^^^^^^^^^^^^^^^^^^^^^^
+
+You have two options to compile your model:
+1. Compile your model within ``model_build_fn`` and return this
+compiled model. In this case, SciKeras will not re-compile your model
+and all compilation parameters (such as ``optimizer``) given to
+:py:func:`scikeras.wrappers.BaseWrapper.__init__` will be ignored.
```

Comment:
> I think we should check to make sure the model is compiled correctly.

I'm not sure what you mean by compiled correctly? If `compile` was run and it did not produce any errors, then the model is compiled correctly, right?

> changing parameters won't affect the behavior of `BaseWrapper`

If you mean the case where the user compiles the model within `model_build_fn` but also passes an `optimizer != None` then I would expect that that `optimizer` is ignored (perhaps with a warning).

Reply:
> If you mean the case where the user compiles the model within `model_build_fn` but also passes an `optimizer != None` then I would expect that that optimizer is ignored (perhaps with a warning).

That's the case I'm talking about. Why are you so opposed to making sure the parameters given to `BaseWrapper.__init__` match those of the returned model? A fair amount of the `BaseWrapper.__init__` parameters are None (5 out of 10), include `loss` and `batch_size`.

Reply:
I am opposed because:
1. doing those checks would be extremely complex. 
2. I think that if users want to compile their own models, we should just let them do that. At most, we can issue a warning.

Reply:
At the very least, I think warnings should be provided. I think these warnings should be tested.

This is work for a future PR. I've added it to #68.

Reply:
Warnings would be ~easyish. Just check if any of these parameters are != their default value (or if any new routed parameters were passed) and issue a warning. This would result in some extra warnings, but at least it's a clear explicit check.